packages/x-date-pickers/src/internals/hooks/usePickerState.ts

```diff
@@ -69,99 +151,98 @@ export const usePickerState = <TInput, TDateValue>(
     }
   }, [parsedDateValue]);
 
-  const [draftState, dispatch] = React.useReducer(
-    (state: Draftable<TDateValue>, action: DraftAction<TDateValue>): Draftable<TDateValue> => {
-      switch (action.type) {
-        case 'reset':
-          return initDraftableDate(action.payload);
-        case 'update':
-          return {
-            ...state,
-            draft: action.payload,
-          };
-        default:
-          return state;
-      }
-    },
-    parsedDateValue,
-    initDraftableDate,
-  );
-  if (!valueManager.areValuesEqual(utils, draftState.committed, parsedDateValue)) {
-    dispatch({ type: 'reset', payload: parsedDateValue });
-  }
-
-  const [initialDate, setInitialDate] = React.useState<TDateValue>(draftState.committed);
-
-  // Mobile keyboard view is a special case.
-  // When it's open picker should work like closed, because we are just showing text field
-  const [isMobileKeyboardViewOpen, setMobileKeyboardViewOpen] = React.useState(false);
+  React.useEffect(() => {
+    if (isOpen) {
+      // Update all dates in state to equal the current prop value
+      setDate({ action: 'setAll', value: parsedDateValue, internal: true });
+    }
+  }, [isOpen]); // eslint-disable-line react-hooks/exhaustive-deps
 
-  const acceptDate = React.useCallback(
-    (acceptedDate: TDateValue, needClosePicker: boolean) => {
-      onChange(acceptedDate);
+  // Set the draft and committed date to equal the new prop value.
+  if (!valueManager.areValuesEqual(utils, dateState.committed, parsedDateValue)) {
+    setDate({ action: 'setCommitted', value: parsedDateValue, internal: true });
+  }
 
-      if (needClosePicker) {
-        setIsOpen(false);
-        setInitialDate(acceptedDate);
-        if (onAccept) {
-          onAccept(acceptedDate);
-        }
-      }
-    },
-    [onAccept, onChange, setIsOpen],
-  );
+  const shouldCloseOnSelect = !(disableCloseOnSelect ?? wrapperVariant === 'mobile');
 
-  const wrapperProps = React.useMemo(
+  const wrapperProps = React.useMemo<PrivateWrapperProps>(
     () => ({
       open: isOpen,
-      onClear: () => acceptDate(valueManager.emptyValue, true),
-      onAccept: () => acceptDate(draftState.draft, true),
-      onDismiss: () => acceptDate(initialDate, true),
+      onClear: () => {
+        // Reset all date in state to the empty value and close picker
+        setDate({ value: valueManager.emptyValue, action: 'acceptAndClose' });
+      },
+      onAccept: () => {
+        // Set all date in state to equal the current draft value and close picker
+        setDate({ value: dateState.draft, action: 'acceptAndClose' });
+      },
+      onDismiss: () => {
+        if (shouldCloseOnSelect) {
```

Comment:
Why do you need to check `shouldCloseOnSelect` here? As I understand, if `onDismiss` is called then we need to set the date based on whether it is full or partial.

Reply:
The behavior of `onDismiss` is under discussion
On `master` now we reset the date
But if `shouldCloseOnSelect = false`, then we probably want to `acceptAndClose` because clicking outside of the picker is the main way to validate the value.

Things become more complicated for multi-step input,
If a user sets the date and time, then re-open the picker, select a date and close : should it keep the new date or not ?
Currently it does not, but one might argue that it should.